vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -131,14 +134,78 @@ void clear_pred_and_label(ccb& data)
 // true if there exists at least 1 action in the cb multi-example
 bool has_action(multi_ex& cb_ex) { return cb_ex.size() > 1; }
 
+// This function intentionally does not handle increasing the num_features of the example because
+// the output_example function has special logic to ensure the number of feaures is correctly calculated.
+// Copy anything in default namespace for decision to ccb_decision_namespace in shared
+// Copy other decision namespaces to shared
+void inject_decision_features(example* shared, example* decision)
+{
+  for (auto index : decision->indices)
+  {
+    // constant namespace should be ignored, as it already exists and we don't want to double it up.
+    if (index == constant_namespace)
+    {
+      continue;
+    }
+    else if (index == 32)  // Decision default namespace has a special namespace in shared
+    {
+      LabelDict::add_example_namespace(*shared, (char)ccb_decision_namespace, decision->feature_space[32]);
+    }
+    else
+    {
+      LabelDict::add_example_namespace(*shared, static_cast<char>(index), decision->feature_space[index]);
+    }
+  }
+}
+
+void remove_decision_features(example* shared, example* decision)
+{
+  for (auto index : decision->indices)
+  {
+    // constant namespace should be ignored, as it already exists and we don't want to double it up.
+    if (index == constant_namespace)
+    {
+      continue;
+    }
+    else if (index == 32)  // Decision default namespace has a special namespace in shared
+    {
+      LabelDict::del_example_namespace(*shared, static_cast<char>(ccb_decision_namespace), decision->feature_space[32]);
+    }
+    else
+    {
+      LabelDict::del_example_namespace(*shared, static_cast<char>(index), decision->feature_space[index]);
+    }
+  }
+}
+
+// Generates all pairs of the namespaces for the two examples
+void calculate_and_insert_interactions(example* shared, example* decision, std::vector<std::string>& vec)
```

Comment:
We should reserve size in vec. It's `len(shared->indices)` * `len(decision->indices)` + `len(vec)`